File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import defaultdict

def solution(n, computers):
    route = defaultdict(set)
    for i in range(n):
        for j in range(len(computers[i])):
            if computers[i][j] == 1 and i != j:
                route[i].add(j)
                route[j].add(i)
                   
    uf = [i for i in range(n)]
    
    def find(a):
        if uf[a] != a:  
            uf[a] = find(uf[a])
        return uf[a] 
    
    def union(a,b):
        a = find(a) 
        b = find(b)
        
        if a > b:
            uf[a] = b
        elif b > a:
            uf[b] = a
    
    for i in route.keys():
        for j in route[i]:
            union(i,j)
    
    return len(set(uf))
    
    
    

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    public static int[] uf;
        
    public int solution(int n, int[][] computers) {
        uf = new int[n];
        for(int i = 0; i < n; i++){
            uf[i] = i;
        }
        for(int k = 0; k < 2; k++){
            for(int i = 0; i < n; i ++){
                for(int j = 0; j < n; j ++){
                    if(i != j && computers[i][j] == 1){
                        union(i,j);
                    }
                }
            }
        }
        Set<Integer> ans = new HashSet<>();
        for(Integer i : uf){
            ans.add(i);
        }
        
        return ans.size();
    }
    
    public int find(int a){
        if(uf[a] == a){
            return a;
        }
        uf[a] = find(uf[a]);
        return uf[a];
    } 
    
    public void union(int a, int b){
        a = find(a);
        b = find(b);
        
        if(a > b){
            uf[a] = b;
        }
        else if(b > a){
            uf[b] = a;
        }
    }
        
}